WordPress/gutenberg
# Experiment: Comments block: client-side form submission

## What?



Built on top of https://github.com/WordPress/gutenberg/pull/53733. Supersedes https://github.com/WordPress/gutenberg/pull/49305.

This PR adds an option to the Comments block to replace the current server-side form submission with a client-side one.

## Why?



To improve the user experience.

## How?



The form submission is intercepted using a directive (`data-wp-on--submit`), which stops the regular form behavior and makes an HTTP request using JavaScript. If the submission goes through successfully, the comments list is updated with the HTML response. If the submission fails, the problem is shown on the screen.

## Tasks

There are still many tasks that need to be done before this can be considered ready. I'll probably add more as we progress and find out what needs to be done.

- [x] Add a UI option to the Comments block. This is disabled by default.
- [x] Only add `data-wp-navigation-id` if the option is enabled.
- [x] Pass down that option using Block Context.
- [x] Only add the directives, `supports.interactivity` and enqueue the `view.js` file in the Post Comments Form block if the option is enabled.
- [x] Add `data-wp-key` to the individual comments.
- [x] If something fails in the JS logic, fall back to server-side submission.
- [x] Include the hash with the comment ID in the URL to facilitate linking.
- [x] Scroll to the new comment on success.
- [x] Add loading status while sending the comment.
- [x] Scroll to the error on failure (if it's not already on the screen).
- [x] Add accessibility.
- [x] Improve the error display (styles).
- [x] Move the form when the `reply` buttons are clicked.
- [x] Reset the position of the form when it is successfully submitted.
- [x] Change the hidden input that holds the parent comment id when using the `reply` button.
- [x] Changed the form title to "Reply to Author" when using the `reply` button.
- [x] Show the `cancel reply` buttons when using the `reply` button.
- [x] Control all the inputs so they are not wiped out when the form is moved around.
- [x] Move improved `data-wp-bind` logic to its own PR. https://github.com/WordPress/gutenberg/pull/54003
- [x] Move `data-wp-slot`/`data-wp-fill` implementation to its own PR. https://github.com/WordPress/gutenberg/pull/53958

### Tasks left for subsequent PRs

- Render the error message using `data-wp-dangerous-html` instead of `data-wp-text`.
- Display a popup when a user tries to enable the enhanced submission and the Comments Form block is not included inside the Comments block. The user should be able to choose between canceling the action, or adding the Comments Form block (added at the end) using buttons in the popup.
- Let people configure the position and styles of the error messages using a different block.
- Improve the loading indicator. It'd be better not to change the text of the submit button and use an animation instead.
- Move the error styles to a separate CSS file and only enqueue it when the enhanced submission is activated.
- Don't move the scroll to the elements if they are already inside the viewport.
- Add a `dom` option to `navigate` that accepts the parsed DOM instead of the `html` and use that instead of the `html`.
- Maintain focus and selected text range after moving the form below the comment being replied to.

## Testing Instructions






To test the initial implementation:

- Add or use a template with the Comments block.
- Submit a comment.
- The page should not refresh, and the new comment should appear on the screen.
- If there's an error, you should see the error below the submit button.
- The number of comments should be updated.
- If the comment is awaiting moderation, the message should appear.

### Testing Instructions for Keyboard



None yet.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3305402/bc8a6fdd-2e72-467c-9ef5-da8c99672b81
